proposals/crd-autoscaler.md

```diff
@@ -0,0 +1,167 @@
+  * Scales vLLM pods up or down based on policy-defined thresholds.
+* **Metrics Exporter**: Exposes inference metrics to Kubernetes via Prometheus.
+
+#### Custom Resources
+
+The primary custom resource is the `vLLMScalingPolicy`, defined as follows:
+
+```go
+// vLLMScalingPolicySpec defines the desired state of vLLMScalingPolicy
+type vLLMScalingPolicySpec struct {
+    // ScaleTargetRef points to the scale-able resource that this autoscaler should target, e.g., Deployment.
+    ScaleTargetRef corev1.ObjectReference `json:"scaleTargetRef"`
+
+    // MinReplicas is the minimum number of replicas to which the target can be scaled down.
+    MinReplicas int32 `json:"minReplicas"`
+
+    // MaxReplicas is the maximum number of replicas to which the target can be scaled up.
+    MaxReplicas int32 `json:"maxReplicas"`
+
+    // ScalingRules defines the conditions under which scaling occurs.
+    ScalingRules []ScalingRule `json:"scalingRules"`
+}
+
+// ScalingRule defines an individual rule for autoscaling decisions.
+type ScalingRule struct {
+    // Metric is the key metric used for scaling (e.g., gpu_utilization, queue_length, latency).
+    Metric string `json:"metric"`
+
+    // Threshold is the value at which scaling action should be taken.
+    Threshold int `json:"threshold"`
+}
+```
+
+#### Example CRD YAML
+
+```yaml
+apiVersion: autoscaler.production-stack/v1alpha1
+kind: vLLMScalingPolicy
+metadata:
+  name: example-policy
+spec:
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: meta-llama/Llama-2-70b-hf
+  minReplicas: 1
+  maxReplicas: 10
+  scalingRules:
+    - metric: "gpu_utilization"
+      threshold: 75
+    - metric: "queue_length"
+      threshold: 10
+    - metric: "latency"
+      threshold: 1000
+```
+
+#### Interface Changes
+
+1. **vLLM Helm Chart Modifications**:
+   * Add `autoscaler.enabled` flag.
+   * Expose GPU utilization metrics via Prometheus.
+   * Provide annotation-based discovery for autoscaler.
+
+2. **vLLM Autoscaler API**:
+   * `GetScalingMetrics(pod)` - Fetches real-time inference metrics.
+   * `AdjustReplicas(model, targetReplicas)` - Adjusts vLLM replicas dynamically.
+   * `GetScalingPolicy(model)` - Fetches the scaling policy for a given model.
+
+### Test Plans
+
+* Unit tests for controller logic.
+* End-to-end tests for scaling behavior validation.
+* Performance tests simulating bursty workloads.
+
+## Drawbacks
+
+* **Increased Complexity**: Adds another component to Kubernetes.
+* **Potential Overhead**: Frequent scaling may introduce latency.
+* **Failure Modes**: Scaling decisions must be robust to noisy metrics.
+
+## Alternatives
+
+* **Kubernetes HPA/VPA**: Lacks inference-specific metrics support.
+* **Do Nothing**: Requires manual adjustments.
+
+## Implementation Timeline / Phases
+
+### Phase 1: Core Infrastructure
+
+* Implement `vLLMScalingPolicy` CRD.
+* Develop the autoscaler controller.
+* Integrate Prometheus metrics exporter.
```

Comment:
I'm curious to know why not re-using KEDA and the prometheus scaler https://keda.sh/docs/2.16/scalers/prometheus/ ?